website/content/development/gpio.md

```diff
@@ -85,3 +84,101 @@ At this point, we should be able to see the high/low signal on the multimeter,
 alternating between 3.3V (high) and 0V (low) every 5 seconds:
 
 <a href="/img/2020-06-15-gpio.jpg"><img src="/img/2020-06-15-gpio.thumb.jpg" srcset="/img/2020-06-15-gpio.thumb.2x.jpg 2x,/img/2020-06-15-gpio.thumb.3x.jpg 3x" width="700" style="border: 1px solid grey; margin-bottom: 2em; margin-top: 1em"></a>
+
+## Setting up an input pin
+
+To setup a Pin as an input and to act on change to High level, such as a button,
+we must poll the Pin periodically. This is needed because the
+[periph.io/device/button](https://periph.io/device/button/) example relies on
+the fact that the Raspberry Pi OS uses a custom kernel driver that enables the
+sysfs GPIO interface, which is not available on the upstream kernel.
+
+The below example configures a single Pin for Input and uses Go channels to
+signal level changes that can be acted on. Keep in mind that shortening the
+polling interval will increase CPU usage.
+
+```go
+package main
+
+import (
+	"log"
+	"time"
+
+	"periph.io/x/conn/v3/gpio"
+	"periph.io/x/conn/v3/gpio/gpioreg"
+	host "periph.io/x/host/v3"
+)
+
+type PinLevelMessage struct {
+	State gpio.Level
+	Reset gpio.Level
+}
+
+func setupGPIOInput(pinName string, levelChan chan PinLevelMessage) (gpio.PinIO, error) {
+	log.Printf("Loading periph.io drivers")
+	if _, err := host.Init(); err != nil {
+		return nil, err
+	}
+
+	// Find Pin by name
+	p := gpioreg.ByName(pinName)
+
+	// Configure Pin for input, with an internal pull down resistor.
+	// Edge mode is currently not supported
+	if err := p.In(gpio.PullDown, gpio.NoEdge); err != nil {
+		return nil, err
+	}
+
+	// Setup Input signalling
+	go func() {
+		lastLevel := p.Read()
+		// How often to poll levels, 100-150ms is fairly responsive unless
+		// button presses are very fast.
+		// Shortening the polling interval <100ms significantly increases
+		// CPU load.
+		t := time.NewTicker(125 * time.Millisecond)
+		for {
+			currentLevel := p.Read()
+			log.Printf("level: %v", currentLevel)
+
+			if currentLevel != lastLevel {
+				levelChan <- PinLevelMessage{State: currentLevel, Reset: !currentLevel}
+				lastLevel = currentLevel
+			}
+			<-t.C
+		}
+	}()
+	return p, nil
+}
+
+func main() {
+	// Channel for communicating Pin levels
+	levelChan := make(chan PinLevelMessage)
+
+	p, err := setupGPIOInput("GPIO4", levelChan)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Main loop, act on level changes
+	for {
+		// How often to check for channel changes, generally it is fine to have
+		// a slightly longer interval here because the signal change is already
+		// caught at this point and will be waiting in the channel.
+		t := time.NewTicker(200 * time.Millisecond)
```

Comment:
Why this additional level of delay? You could just read from levelChan blockingly, no?

Reply:
Not sure why I put this in here.. possible before I could get the blocking channel to work 🤔. I've removed it and tested it w/o, works fine!